kernel/local_storage_initializer/src/lib.rs

```diff
@@ -329,64 +393,132 @@ impl TlsInitializer {
-            TlsDataImage {
-                _data: Some(data_copy),
-                ptr:   tls_self_ptr_value,
+            LocalStorageDataImage {
+                ptr: tls_self_ptr_value,
+                data: data_copy,
+                _phantom: PhantomData,
             }
         }
 
         #[cfg(target_arch = "aarch64")] {
-            let data_copy: Box<[u8]> = self.data_cache.as_slice().into();
-            let ptr = data_copy.as_ptr() as usize;
-            TlsDataImage {
-                _data: Some(data_copy),
-                ptr,
+            let cloned = self.data_cache.clone();
+            LocalStorageDataImage {
+                ptr: cloned.as_ptr() as u64,
+                data: cloned,
+                _phantom: PhantomData,
             }
         }
     }
 }
 
-/// An initialized TLS area data image ready to be used by a new task.
+/// An initialized data image ready to be used by a CPU/new task.
 /// 
-/// The data is opaque, but one can obtain a pointer to the TLS area.
-/// 
-/// The enclosed opaque data is stored as a boxed slice (`Box<[u8]>`)
-/// instead of a vector (`Vec<u8>`) because it is instantiated once upon task creation
-/// and should never be expanded or shrunk.
+/// The data is opaque, but one can obtain a pointer to the CLS/TLS area.
 /// 
 /// The data is "immutable" with respect to Theseus task management functions
-/// at the language level.
-/// However, the data within this TLS area will be modified directly by code
-/// that executes "in" this task, e.g., instructions that access the current TLS area.
+/// at the language level. However, it will be modified by CLS/TLS accesses.
 #[derive(Debug)]
-pub struct TlsDataImage {
-    // The data is wrapped in an Option to avoid allocating an empty boxed slice
-    // when there are no TLS data sections.
-    _data: Option<Box<[u8]>>,
-    ptr:   usize,
+pub struct LocalStorageDataImage<T>
+where
+    T: LocalStorage,
+{
+    data: Vec<u8>,
+    ptr: u64,
+    _phantom: PhantomData<T>,
 }
-impl TlsDataImage {
-    /// Creates an empty TLS data image with no TLS section content.
-    pub const fn empty() -> TlsDataImage {
-        Self { _data: None, ptr: 0 }
+
+impl<T> LocalStorageDataImage<T>
+where
+    T: LocalStorage,
+{
+    /// Creates an empty data image.
+    pub const fn new() -> Self {
+        Self {
+            data: Vec::new(),
+            ptr: 0,
+            _phantom: PhantomData,
+        }
     }
 
-    /// Sets the current CPU's TLS register to point to this TLS data image.
+    /// Inherits the data from another data image.
+    ///
+    /// # Panics
     ///
-    /// * On x86_64, this writes to the `FsBase` MSR.
-    /// * On ARMv8, this writes to `TPIDR_EL0`.
-    pub fn set_as_current_tls_base(&self) {
+    /// Panics if the other image has a longer length, or (on x64) if the other
+    /// image has a differently sized static area.
+    pub fn inherit(&mut self, other: &Self) {
```

Comment:
perhaps i missed something, but I don't see where this is used or why we need it. What's the intended use of this?

Reply:
we use it in `reload_current_core`.

Reply:
I see, thanks. This seems a little bit overly complex; I'm not sure why we need the extra complexity as compared to TLS (aside from, of course, needing to swap the new CLS data region into the list of all CLS data regions.

But no real need to change it, if it works it's fine.

Reply:
Well because we are appending new data onto the active CLS image. The old part of the CLS image has already been used, so its data will be in some different state that we need to copy over into the new image.

Reply:
Right, so I think TLS is similar but simpler overall. The (old) TLS initializer created and maintained a single global image, which it then hands out cloned instances of to each new task.

But yes, in this case, we do need to preserve the modified data on a per-CPU basis, which is an additional requirement beyond TLS. TLS doesn't need that due to the fact that each new task will only be created _after_ its crate (and its dependencies) have been linked in, which includes all of its TLS sections being loaded. There's no case (yet) where a task would suddenly need to expand it TLS data area in order to access sections that weren't loaded at the time when it initially began executing.

I think my original comment came from my old assumption/mental model about how CLS worked, from the old days of the `PerCpuData` struct that was stored in a dedicated 1-page `MappedPages` for each CPU -- I was expecting to see a single (or two) pages per CPU that stored its CLS data, and would only permit growing (not removing sections). But that approach isn't unlimited since it's restricted by the remaining space available in the page(s).